python/delta_sharing/rest_client.py

```diff
@@ -143,16 +143,63 @@ def __init__(self, profile: DeltaSharingProfile, num_retries=10):
         self._profile = profile
         self._num_retries = num_retries
         self._sleeper = lambda sleep_ms: time.sleep(sleep_ms / 1000)
+        self.auth_session(profile)
 
+    def auth_session(self, profile):
         self._session = requests.Session()
+        self.auth_broker(profile)
+        if urlparse(profile.endpoint).hostname == "localhost":
+            self._session.verify = False
+
+    def auth_broker(self, profile):
+        if profile.share_credentials_version == 2:
+            if profile.type == "persistent_oauth2.0":
+                self.auth_persistent_oauth2(profile)
+            elif profile.type == "bearer_token":
+                self.auth_bearer_token(profile)
+            elif profile.type == "basic":
+                self.auth_basic(profile)
+            else:
+                self.auth_bearer_token(profile)
+        else:
+            self.auth_bearer_token(profile)
+
+    def auth_bearer_token(self, profile):
         self._session.headers.update(
             {
                 "Authorization": f"Bearer {profile.bearer_token}",
                 "User-Agent": DataSharingRestClient.USER_AGENT,
             }
         )
-        if urlparse(profile.endpoint).hostname == "localhost":
-            self._session.verify = False
+
+    def auth_persistent_oauth2(self, profile):
+        response = requests.post(profile.token_endpoint,
+                                 data={"grant_type": "client_credentials"},
+                                 auth=(profile.client_id,
+                                       profile.client_secret),)
+        bearer_token = "{}".format(response.json()["access_token"])
+
+        self._session.headers.update(
+            {
+                "Authorization": f"Bearer {bearer_token}",
+                "Content-Type": "application/x-www-form-urlencoded",
+                "Accept": "application/json",
+                "User-Agent": DataSharingRestClient.USER_AGENT,
+            }
+        )
+
+    def auth_basic(self, profile):
+        response = requests.post(profile.endpoint,
```

Comment:
This does not look right: you are making a request to the API endpoint to request a bearer token. Instead, the username and password should be used as the credential to talk to the API endpoint directly.